### DISCUSSION TITLE
HttpClient throws exception with 'Error decoding HTTP response body: Error decoding JSON stream for type [list]' -- how do I fix it?

### DISCUSSION THREAD

**Discussion by nedelva** [micronaut-projects/micronaut-core#10915] [Q&A] [answered]
I have this thin http client wrapper over a REST service:
```java
@Singleton
@Slf4j
public class TransversalApiLowLevelClient {
    private final HttpClient httpClient;
    public TransversalApiLowLevelClient(@Client(id = "transversal-api") HttpClient httpClient, TransversalApiConfiguration configuration) {
        this.httpClient = httpClient;
       // other initializations
    }

    public Mono<List<Auto>> getAutoContractsAsync(String enterpriseId) {
        //fetch the data from somewhere..
        String endpointUrl // = ... ; 
        MutableHttpRequest<?> req = HttpRequest.GET(endpointUrl).accept(MediaType.APPLICATION_JSON_TYPE);
        Publisher<List<Auto>> publisher = httpClient.retrieve(req, Argument.listOf(Auto.class));
        return Mono.from(publisher)
                .defaultIfEmpty(Collections.emptyList());
    }
```

Then in the test I want to test the case where the service returns "no data" :

```java
    @Test
    @SneakyThrows
    void getAutoContractsAsyncWhenEmpty() {
        wmExtension.stubFor(get(urlMatching("/some-path-here/auto[?\\w%&=_,]*$"))
                .willReturn(okForJson("[]")));
        Mono<List<Auto>> autoContractsAsync = transversalApiLowLevelClient.getAutoContractsAsync("0200068636");
        List<Auto> autoList = autoContractsAsync.defaultIfEmpty(Collections.emptyList()).block();
        assertThat(autoList.size(), CoreMatchers.is(0));
    }
```
`wmExtension ` is defined as a static class variable:
```java
@RegisterExtension
static WireMockExtension wmExtension = WireMockExtension.newInstance()
        .options(wireMockConfig().port(4444).dynamicHttpsPort())
        .build();
```

Problem is,  running the above test throws an error:

> 11:45:38.579 [other-nioEventLoopGroup-1-1] DEBUG io.micronaut.http.client.netty.DefaultHttpClient - Sending HTTP GET to http://localhost:4444/some-path-here/auto?Enterprise_Nbr=0200068636
> 
> io.micronaut.http.client.exceptions.HttpClientResponseException: Client 'transversal-api': Error decoding HTTP response body: Error decoding JSON stream for type [list]: Cannot construct instance of `com.acme.data.model.Auto` (although at least one Creator exists): no String-argument constructor/factory method to deserialize from String value ('[]')
>  at [Source: REDACTED (`StreamReadFeature.INCLUDE_SOURCE_IN_LOCATION` disabled); line: 1, column: 1] (through reference chain: java.lang.Object[0])
> .....
> Caused by: io.micronaut.http.codec.CodecException: Error decoding JSON stream for type [list]: Cannot construct instance of `com.acme.data.model.Auto` (although at least one Creator exists): no String-argument constructor/factory method to deserialize from String value ('[]')
>  at [Source: REDACTED (`StreamReadFeature.INCLUDE_SOURCE_IN_LOCATION` disabled); line: 1, column: 1] (through reference chain: java.lang.Object[0])
> 	at io.micronaut.json.body.JsonMessageHandler.decorateRead(JsonMessageHandler.java:94)
> 	at io.micronaut.json.body.JsonMessageHandler.read(JsonMessageHandler.java:108)
> 	at io.micronaut.http.netty.body.NettyJsonHandler.read(NettyJsonHandler.java:129)
> 	at io.micronaut.http.client.netty.FullNettyClientHttpResponse.convertByteBuf(FullNettyClientHttpResponse.java:224)
> ....
> Caused by: com.fasterxml.jackson.databind.exc.MismatchedInputException: Cannot construct instance of `com.acme.data.model.Auto` (although at least one Creator exists): no String-argument constructor/factory method to deserialize from String value ('[]')
>  at [Source: REDACTED (`StreamReadFeature.INCLUDE_SOURCE_IN_LOCATION` disabled); line: 1, column: 1] (through reference chain: java.lang.Object[0])
> 	at com.fasterxml.jackson.databind.exc.MismatchedInputException.from(MismatchedInputException.java:63)
> 	at com.fasterxml.jackson.databind.DeserializationContext.reportInputMismatch(DeserializationContext.java:1754)
> 	at com.fasterxml.jackson.databind.DeserializationContext.handleMissingInstantiator(DeserializationContext.java:1379)
> 	at com.fasterxml.jackson.databind.deser.std.StdDeserializer._deserializeFromString(StdDeserializer.java:311)
> 	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromString(BeanDeserializerBase.java:1588)
> 	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeOther(BeanDeserializer.java:197)
> 	at com.fasterxml.jackson.module.afterburner.deser.SuperSonicBeanDeserializer.deserialize(SuperSonicBeanDeserializer.java:120)
> 	at com.fasterxml.jackson.databind.deser.std.CollectionDeserializer.handleNonArray(CollectionDeserializer.java:417)

The project uses Micronaut v4.4.3 with Jackson Databind library (v4.4.10)

For the sake of completeness, the following test completes normally (when the service returns some records):
```java
    @Test
    @SneakyThrows
    void getAutoContractsAsync() {
        String jsonBody = IO.loadFileResource("auto_contracts_0200068636.json"); // loads a canned response from a local file
        wmExtension.stubFor(get(urlMatching("/some-path-here/auto[?\\w%&=_,]*$"))
                .willReturn(aResponse()
                        .withHeader("Content-Type", "application/json")
                        .withBody(jsonBody)));
        Mono<List<Auto>> autoContractsAsync = transversalApiLowLevelClient.getAutoContractsAsync("0200068636");
        List<Auto> autoList = autoContractsAsync.block();
        assertThat(autoList.size(), CoreMatchers.is(7));
    }
```

So how do I fix the "no-data" case? 

AFAIK, _in general_ a service can either

1. Return an empty body with a 204 http response code, or
2. Return a "[]" as body with a 200 http response code

when there is no data available.

**Comment by nedelva** [comment]
Could it be related with [this issue](https://github.com/micronaut-projects/micronaut-core/issues/10810)?

**Comment by yawkat** [comment, marked as answer]
Are you sure that `okForJson("[]")` returns `[]` and not `"[]"`

**Reply by nedelva** [reply]
Thank you @yawkat for this question! You made me look into the matter more closely and it turns out I have been using the WireMock API wrongly. The right call to made is not `okForJson("[]")` but `okForJson(new ArrayList<Auto>())` since it expects an object, not the raw JSON string.
Here's the definition of that method:
```java
public static <T> ResponseDefinitionBuilder okForJson(T body) {
    return responseDefinition().withStatus(200).withBody(Json.write(body)).withHeader("Content-Type", "application/json");
}
```

One other thing to note is that is crucial to set the "content-type" header. When I tried to use `.willReturn(aResponse().withBody("[]"))`, I ran into an exception.
So a set up like this one
```java
wmExtension.stubFor(get(urlMatching("/some-path-here/auto[?\\w%&=_,]*$"))
        .willReturn(aResponse().withBody("[]")));
```

will make the http client throw an exception, while this goes smoothly:

```java
wmExtension.stubFor(get(urlMatching("/some-path-here/auto[?\\w%&=_,]*$"))
        .willReturn(aResponse().withHeader("Content-Type", "application/json").withBody("[]")));
```
